### DISCUSSION TITLE
Ensure middleware runs on every next/link navigation

### DISCUSSION THREAD

**Discussion by joulev** [vercel/next.js#34128] [Help]
#### `pages/_middleware.js`

```js
import { NextResponse } from "next/server";

export function middleware(req) {
    console.log("A request has been made to", req.nextUrl.pathname);
    return NextResponse.next();
}
```

#### `pages/index.jsx`

```jsx
import Link from "next/link";

export default function Index() {
  return (
    <Link href="/test">Test</Link>
  );
}
```

#### `pages/test.jsx`

```jsx
import Link from "next/link";

export default function Index() {
  return (
    <Link href="/">Back</Link>
  );
}
```

If I build in production mode (`npm run build` then `npm start`), then go to the page and continuously clicking the links, the log is only printed three times.

```
A request has been made to /
A request has been made to /test
A request has been made to /
```

All subsequent navigation (can I even call them request?) didn't call the middleware function.

How can I ensure that the middleware function is ran on every request and every nagivation using `next/link` (or `next/router` for that matter)?

**Comment by btbenedi** [comment]
This would be extremely helpful, we even tried to use the below to no avail

`<Link href="/test" prefetch={false}>Test</Link>`

**Comment by Andrewjeska** [comment]
What's the correct way to use middleware for auth checks then? I think there's some sort of caching going on, but I'd have to div e into the codebase

**Comment by Andrewjeska** [comment]
`
res.headers.set(`x-middleware-cache`, `no-cache`);
`
See:
https://github.com/vercel/next.js/issues/30938#issuecomment-970239412
https://stackoverflow.com/a/73096326

**Reply by hjaber** [reply]
> ` res.headers.set(`x-middleware-cache`, `no-cache`); ` See: [#30938 (comment)](https://github.com/vercel/next.js/issues/30938#issuecomment-970239412) https://stackoverflow.com/a/73096326

With next.js 13 & app directory, I find middleware running only once despite setting the response with `("x-middleware-cache", "no-cache")`

Are you finding that middleware runs on every request?
